lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,817 @@
+/* -*- mode: c++; c-basic-offset: 2 -*- */
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "resolve-labels.h"
+#include "../parser/message.h"
+#include "../parser/parse-tree-visitor.h"
+#include <cstdarg>
+#include <iostream>
+#include <cctype>
+#include <cassert>
+
+namespace {
+
+using namespace Fortran;
+using namespace parser::literals;
+using ParseTree_t = parser::Program;
+using CookedSource_t = parser::CookedSource;
+using Index_t = parser::CharBlock;
+using IndexList = std::vector<std::pair<Index_t, Index_t>>;
+using Scope_t = unsigned;
+using LblStmt_t = std::tuple<Scope_t, Index_t, unsigned>;
+using ArcTrgt_t = std::map<parser::Label, LblStmt_t>;
+using ArcBase_t = std::vector<std::tuple<parser::Label, Scope_t, Index_t>>;
+
+const bool StrictF18 = false; // FIXME - make a command-line option
+
+const unsigned DO_TERM_FLAG = 1u;
+const unsigned BRANCH_TARGET_FLAG = 2u;
+const unsigned FORMAT_STMT_FLAG = 4u;
+
+// convenient package for error reporting
+struct ErrorHandler {
+public:
+  explicit ErrorHandler(const parser::CookedSource& CookedSource)
+    : cookedSource{CookedSource}, messages{parser::Messages()} {}
+  ~ErrorHandler() = default;
+  ErrorHandler(ErrorHandler&&) = default;
+  ErrorHandler() = delete;
+  ErrorHandler(const ErrorHandler&) = delete;
+  ErrorHandler& operator=(const ErrorHandler&) = delete;
+
+  parser::Message& Report(const parser::CharBlock& CB,
+			  const parser::MessageFixedText& Fixed, ...) {
+    va_list ap;
+    va_start(ap, Fixed);
+    parser::MessageFormattedText Msg{Fixed, ap};
+    va_end(ap);
+    return messages.Say(parser::Message{CB, Msg});
+  }
+
+  const parser::CookedSource& cookedSource;
+  parser::Messages messages;
+};
+
+/// \brief Is this a legal DO terminator?
+/// Pattern match dependent on the standard we're enforcing
+/// F18:R1131 (must be CONTINUE or END DO)
+template<typename A> constexpr bool IsLegalDoTerm(const parser::Statement<A>&) {
+  using EDS = parser::EndDoStmt;
+  return std::disjunction<std::is_same<A,common::Indirection<EDS>>,
+			  std::is_same<A,EDS>>::value;
+}
+template<> constexpr bool IsLegalDoTerm(const parser::Statement<parser::
+					ActionStmt>& A) {
+  if (std::get_if<parser::ContinueStmt>(&A.statement.u)) {
```

Comment:
Use `std::holds_alternative<>()` rather than `get_if` and an implicit null pointer test.